product-service/src/main/java/org/arpha/service/ProductServiceImpl.java

```diff
@@ -0,0 +1,115 @@
+import org.arpha.dto.product.response.ProductResponse;
+import org.arpha.exception.CreateProductException;
+import org.arpha.exception.ProductNotFoundException;
+import org.arpha.mapper.ProductMapper;
+import org.arpha.mapper.helper.ProductMapperHelper;
+import org.arpha.repository.ProductRepository;
+import org.arpha.utils.Boxed;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+
+import java.util.Set;
+
+@Service
+@RequiredArgsConstructor
+@Transactional
+public class ProductServiceImpl implements ProductService {
+
+    public static final String PRODUCT_NOT_FOUND_MESSAGE = "Product with %d id not found!";
+
+    private final ProductRepository productRepository;
+    private final ProductMapper productMapper;
+    private final ProductMapperHelper productMapperHelper;
+
+    @Override
+    public ProductResponse createProduct(CreateProductRequest createProductRequest) {
+        return Boxed
+                .of(createProductRequest)
+                .mapToBoxed(productMapper::toProduct)
+                .mapToBoxed(productRepository::save)
+                .mapToBoxed(productMapper::toProductResponse)
+                .orElseThrow(() -> new CreateProductException("Error happened during product creation"));
+    }
+
+    @Override
+    public void deleteProduct(long id) {
+       productRepository.deleteById(id);
+    }
+
+    @Override
+    public ProductResponse findProductById(long id) {
+        return Boxed
+                .of(id)
+                .flatOpt(productRepository::findById)
+                .mapToBoxed(productMapper::toProductResponse)
+                .orElseThrow(() -> new ProductNotFoundException(PRODUCT_NOT_FOUND_MESSAGE.formatted(id)));
+    }
+
+    @Override
+    public Page<ProductResponse> findAllProducts(Predicate predicate, Pageable pageable) {
+        return productRepository.findAll(predicate, pageable).map(productMapper::toProductResponse);
+    }
+
+    @Override
+    public ProductResponse findAdminProductById(long id) {
+        return Boxed
+                .of(id)
+                .flatOpt(productRepository::findById)
+                .mapToBoxed(productMapper::toAdminProductResponse)
+                .orElseThrow(() -> new ProductNotFoundException(PRODUCT_NOT_FOUND_MESSAGE.formatted(id)));
+    }
+
+    @Override
+    public Page<ProductResponse> findAdminAllProducts(Predicate predicate, Pageable pageable) {
+        return productRepository.findAll(predicate, pageable).map(productMapper::toAdminProductResponse);
+    }
+
+    @Override
+    public ProductResponse addGenre(long id, Set<String> genres) {
+        return Boxed
+                .of(id)
+                .flatOpt(productRepository::findById)
+                .mapToBoxed(product -> {
+                    product.getGenres().addAll(productMapperHelper.toGenres(genres));
+                    return product;})
```

Comment:
Can you please move it to a separate method. Also we can do this addAll() operation without returning product. Use something like .peek() etc. (if it is available in Boxed)